File: problems/data-structures-n-algorithms/hard/128.js
// 128. Longest Consecutive Sequence
// https://leetcode.com/problems/longest-consecutive-sequence/

const longestConsecutive = function (nums) {
	const map = {};
	let max = 0;

	for (const item of nums) {
		map[item] = 0;
	}

	for (let item of nums) {
		if (map[item] !== 0) {
			continue;
		}

		const _item = item;

		let count = 0;
		while (map[item] === 0 && map[item] !== undefined) {
			map[item] = 1;
			count++;
			item++;
		}

		map[_item] = count + (map[item] || 0);

		if (max < map[_item]) {
			max = map[_item];
		}
	}

	return max;
};

function main() {
	debugger;
	const res = longestConsecutive([100, 4, 200, 1, 3, 2]);
	// const res = longestConsecutive([0, 3, 7, 2, 5, 8, 4, 6, 0, 1]);
	// const res = longestConsecutive([0, 0]);
	// const res = longestConsecutive([0, -1]);

	console.log(res);
}

main();

File: problems/data-structures-n-algorithms/hard/128.cpp
// 128. Longest Consecutive Sequence
// https://leetcode.com/problems/longest-consecutive-sequence/

#include <bits/stdc++.h>
#include <iostream>

using namespace std;

int longestConsecutive(vector<int> &nums) {
	map<int, int> mp;
	int max = 0;

	for (int i = 0; i < nums.size(); i++) {
		mp.insert({nums[i], 0});
	}

	for (int i = 0; i < nums.size(); i++) {
		int item = nums[i];
		int _item = item;

		auto _iter = mp.find(_item);
		auto iter = mp.find(item);
		if (iter->second != 0) {
			continue;
		}

		int count = 0;
		while (iter->second == 0 && iter != mp.end()) {
			iter->second = 1;
			count++;
			item++;
			iter = mp.find(item);
		}

		_iter->second = count + (iter == mp.end() ? 0 : iter->second);

		if (max < _iter->second) {
			max = _iter->second;
		}
	}

	return max;
}

int main() {
	// vector<int> arr = {100, 4, 200, 1, 3, 2};
	// vector<int> arr = {0, 3, 7, 2, 5, 8, 4, 6, 0, 1};
	// vector<int> arr = {0, 0};
	vector<int> arr = {0, -1};
	int res = longestConsecutive(arr);
	cout << res;
	return 0;
}